[WIP] Type annotation improvements

neiljp opened the pull request:
Previous PRs have included mypy annotations but only adapting the code being committed to satisfy travis; these commits contain changes mostly suggested by running mypy locally, now my Debian upgrade makes this much easier.

I'm not sure if there is a consensus of working towards adding Optional, @timabbott ?

(I also found my first bug from applying mypy, strangely enough in bot-related code!)

smarx commented:
*Automated message from Dropbox CLA bot*

@neiljp, it looks like you've already signed the Dropbox CLA. Thanks!

rht commented on api/bots_api/bot_lib.py:
```diff
@@ -51,7 +51,7 @@ def show_error_and_exit(self):
 
 def send_reply(message, response, email, send_message):
-    # type: (Dict[str, Any], str, str, Any) -> Dict[str, Any]
+    # type: (Dict[str, Any], str, str, Callable[[Dict[str, Any]], Dict[str, Any]]) -> Dict[str, Any]
```
Hmmm... Is it possible to run Mypy's type inference on a code to automatically do the annotation? `autopep484`?

neiljp replied:
I've not run mypy myself at this time due to local development limitations, but I could see that it either is or should be possible, since at least the inferred types can be exposed via reveal_type(). That said, if `Any` is marked, mypy considers it as an explicit mark of 'dynamic' code, so presumably wouldn't pay attention in a normal validation run.

This is more of a general question than specifically related to this code; so perhaps ask on [#general](https://chat.zulip.org/#narrow/stream/general)?

timabbott commented:
Also it looks like you're getting mypy errors on the last commit.

neiljp commented:
How does this look now?

Changes:
* Amended commit messages;
* Removed first commit, which is now in the api repo;
* Added Optional to last commit, so Travis shouldn't fail now.

Should mypy check that the types (like Optional) are available?

neiljp commented:
The reported mypy fail is not related to these commits?

timabbott commented:
Yeah, I briefly caused some mypy issues in master on py3.

timabbott commented:
I merged all but the last commit.  For that one, I'm not sure that's the correct solution to the conflict.  I think the tests that are passing `None` are buggy and should be passing plausible request data instead.

I wonder if something is wrong with `tools/run-mypy` itself, since I don't see why these errors weren't being caught by Travis...  are you testing with `./tools/run-mypy --strict-optional`?  Cleaning out the errors we get for that is a project goal.

neiljp commented:
Thanks for the review; the last was the one I wasn't sure about, and yes it was from --strict-optional. As per this one, should tests be adjusted in general? Or based on checks in the functions themselves for equality to None?

timabbott commented:
Yeah, we should change those tests.  In general, it depends -- I'd read the code and see whether the tests are broken.  In general, I think we don't want to be changing the annotations of runtime code to support tests passing in invalid arguments, unless those arguments are actually possible in the real world :).  I guess we can close this PR since all the changes that are useful are merged, and you can just fix those in your next PR (will declutter the project a bit to close this).
